### PR TITLE
Frontend: Avoid mixed content warnings by fixing URLs to use HTTPS

### LINKED ISSUE #4928: HTTP video sources despite HTTPS available
## Bug Description

Stories can be rendered with HTTP video sources despite HTTPS support on the domain, which causes the stories to fail AMP validation.

Example: [this story](https://thetinyherbivore.com/web-stories/fall-outfits-for-easy-hikes/) contains a [HTTP video](http://thetinyherbivore.com/wp-content/uploads/2020/10/Outfit-1-4.mp4) > `source[src]`, but the domain [supports HTTPS](https://thetinyherbivore.com/web-stories/fall-outfits-for-easy-hikes/). See [validation errors](https://validator.amp.dev/#url=https%3A%2F%2Fthetinyherbivore.com%2Fweb-stories%2Ffall-outfits-for-easy-hikes%2F).

## Expected Behavior

- Render story URLs as HTTPS if supported by the domain. 
- If the domain doesn't support HTTPS, display a warning (perhaps in the prepublish checklist).

## Steps to Reproduce

Perhaps WP admin is used in HTTP mode? 

TODO(choumx): Follow up with creator.

## Screenshots

N/A

## Additional Context

- Plugin Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Change all references to home url in WordPress in web stories content to the current schema. 

## Relevant Technical Choices

All references to home url to current schema. Set the value on the model so we dont have to do it everywhere. 

## Testing Instructions

---

Fixes #4928

### CODE REVIEW

**Comment by swissspidy** [comment]
Feels like this should rather be done on the AMP sanitization level / in Story_Renderer, not in the Model class.

**Comment by spacedmonkey** [comment]
In early version of this, add more of the attributes on the story model replaced with https versions. But I scaled it back. So moving this makes sense.

File: includes/Story_Renderer/HTML.php [hunk 1]
```diff
@@ -215,6 +216,23 @@ protected function replace_html_head( $content ) {
...
+	 * @param string $content String to replace.
+	 *
+	 * @return string
+	 */
+	protected function replace_url_scheme( $content ) {
+		$current = is_ssl() ? 'https' : 'http';
+		$new     = is_ssl() ? 'http' : 'https';
+		$home    = get_home_url();
+
+		$content = str_replace( set_url_scheme( $home, $new ), set_url_scheme( $home, $current ), $content );
```
Comment (swissspidy) [inline includes/Story_Renderer/HTML.php]: I had a hard time fully grasping what these variables mean to achieve due to unclear wording. Something like `$current` and `$actual` might be better suited. Can we make this a bit easier to read?

Maybe something like this:

```suggestion
		$home = get_home_url();

		if ( is_ssl() ) {
			// If serving via HTTPS, all content should be referenced via HTTPS.
			return str_replace( set_url_scheme( $home, 'http' ), set_url_scheme( $home, 'https' ), $content );
		}

		// TODO: Is this really needed? Serving assets over HTTPS when accessing story over HTTP is not problematic.
		// We could just do `return $content;`
		return str_replace( set_url_scheme( $home, 'https' ), set_url_scheme( $home, 'http' ), $content );

```

Comment (spacedmonkey) [reply includes/Story_Renderer/HTML.php]: I don't think this is a good change. I think we should improve the comment. Making the code longer and more to excute is not the right course. If it takes a minute to understand, that is a failure in comments and variable names, not the code.

Comment (swissspidy) [reply includes/Story_Renderer/HTML.php]: Even with the latest changes it's not super obvious IMO. Don't get me wrong, the code is super clean and DRY, but IMO it's not as easy to comprehend as something simple as 

```php
if ( is_ssl() ) {
	// If serving via HTTPS, all content should be referenced via HTTPS.
	return str_replace( set_url_scheme( home_url(), 'http' ), set_url_scheme( home_url(), 'https' ), $content );
}
```

just my two cents :)

File: tests/phpunit/tests/Story_Renderer/HTML.php [hunk 2]
```diff
@@ -226,6 +229,33 @@ public function test_sanitizes_and_optimizes_markup() {
...
+
+		$result = $this->call_private_method( $renderer, 'replace_url_scheme', [ set_url_scheme( $link, 'https' ) ] );
+		$this->assertEquals( $result, $link );
+	}
+
+
+	/**
+	 * @covers ::replace_url_scheme
+	 */
+	public function test_replace_url_scheme_invalid_url() {
```
Comment (swissspidy) [inline tests/phpunit/tests/Story_Renderer/HTML.php]: `invalid` doesn't feel like the best term here.

Perhaps something more verbose like this:

```suggestion
	public function test_replace_url_scheme_does_not_change_urls_with_different_host() {
```

or just

```suggestion
	public function test_replace_url_scheme_different_host() {
```

File: includes/Story_Renderer/HTML.php [hunk 3]
```diff
@@ -215,6 +216,23 @@ protected function replace_html_head( $content ) {
...
 
+	/**
+	 * Force home urls to http / https based on context.
+	 *
+	 * @param string $content String to replace.
+	 *
+	 * @return string
+	 */
+	protected function replace_url_scheme( $content ) {
+		$home           = get_home_url();
```
Comment (swissspidy) [inline includes/Story_Renderer/HTML.php]:
```suggestion
		$home           = home_url();
```

With that said, `home_url()` supports a `$scheme` param...

File: includes/Story_Renderer/HTML.php [hunk 4]
```diff
@@ -215,6 +216,23 @@ protected function replace_html_head( $content ) {
...
+		$current_scheme = is_ssl() ? 'https' : 'http';
+		$replace_scheme = is_ssl() ? 'http' : 'https';
+		$search         = set_url_scheme( $home, $replace_scheme );
+		$replace        = set_url_scheme( $home, $current_scheme );
```
Comment (swissspidy) [inline includes/Story_Renderer/HTML.php]:
```suggestion
		$current_scheme = is_ssl() ? 'https' : 'http';
		$replace_scheme = is_ssl() ? 'http' : 'https';
		$search         = home_url( '', $replace_scheme );
		$replace        = home_url( '', $current_scheme );
```

Comment (swissspidy) [reply includes/Story_Renderer/HTML.php]: On that note: confusing that `$replace_scheme` is not related to / used by `$replace`

File: includes/Story_Renderer/HTML.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Story_Renderer/HTML.php]: I tweaked it, just to run on SSL.

File: includes/Story_Renderer/HTML.php (hunk 4 above)
Comment (spacedmonkey) [reply includes/Story_Renderer/HTML.php]: Tweaked.

**Review by swissspidy — APPROVED** [review]

**Pull request merged**